Port to Gtk3 and Python3

Saumya-Mishra9129 opened the pull request:
The activity is working as expected. 
The Pr includes-
Port to Gtk3
Python3
Fixes #2. 
Fixes #3. 
Fixes #7.

Saumya-Mishra9129 committed:
Port to Gtk3 and Python3

Saumya-Mishra9129 committed:
Conflicts resolved

Saumya-Mishra9129 committed:
fix feed_child type-error

chimosky commented on vnclauncher.py:
```diff
@@ -204,9 +204,8 @@ def _configure_vte(self):
-        self.set_colors(Gdk.RGBA.parse(fg_color),
-                        Gdk.RGBA.parse(bg_color),
-                        [])
+        self.set_colors(Gdk.RGBA(*Gdk.color_parse(fg_color).to_floats()),
```
This could be better, you can look at this [implementation](https://github.com/sugarlabs/bichos-activity/blob/f2c5e126c7694ab664d1f5804523c65b7c6b5845/Widgets.py) and decide how you want to proceed.

chimosky commented on vnclauncher.py:
```diff
@@ -133,10 +133,10 @@ def connectVNC(self, button):
             cmd = os.path.join(path, 'x11vnc') + "\n"
             logging.error('Using %s', cmd)
-        self._vte.feed_child(cmd)
+        self._vte.feed_child(cmd.encode(('utf-8')))
```
I'm guessing the double parentheses around `utf-8` is a mistake.

chimosky commented:
Reviewed, thanks.

a1c423b contains changes not included in the commit message.

Starting a server throws this error,
```
/home/ibiam/Activities/VncLauncher/bin/x86-64/x11vnc: error while loading shared
libraries: libssl.so.10: cannot open shared object file: No such file or directory
```

0493182 could have a better commit message.

You can consider demoting logging as the logs seem to be noisy.

As this is a `Port to Gtk3`, you should change the use of `jarabe.model.network` in [L35](https://github.com/sugarlabs/VncLauncher/blob/master/vnclauncher.py#L35) of `vnclauncher.py`, you can look at `sugar-toolkit-gtk3` for a replacement. 

Also consider fixing #7.

Kindly test your changes before making a PR.

Saumya-Mishra9129 committed:
double parentheses typo fixed

chimosky commented on vnclauncher.py:
```diff
@@ -204,9 +204,11 @@ def _configure_vte(self):
-
-        self.set_colors(Gdk.RGBA(*Gdk.color_parse(fg_color).to_floats()),
-                        Gdk.RGBA(*Gdk.color_parse(bg_color).to_floats()), [])
+        f_rgba = Gdk.RGBA()
+        f_rgba.parse(fg_color)
+        b_rgba = Gdk.RGBA()
+        b_rgba.parse(bg_color)
+        self.set_colors(f_rgba, b_rgba, [])
```
You can create just one instance of `Gdk.RGBA` and call that when you need to.

quozl commented:
Please review and re-use any changes from https://github.com/sugarlabs/VncLauncher/pull/4 which was closed after progress stalled.  When re-using, use "git commit --author" to credit properly.

Saumya-Mishra9129 commented:
@quozl Reviewed #4 didn't find anything . The changes are almost same. 
@chimosky 
`/home/ibiam/Activities/VncLauncher/bin/x86-64/x11vnc: error while loading shared
libraries: libssl.so.10: cannot open shared object file: No such file or directory`
can be faced when your SSL is not installed or updated. If libssl is installed then cause of this error is a different name for the same library present. So you have to create a symbolic link with libssl.so.10 to lib.so.1.0.0(if it is the library present in system). Symbolic links can be created by these commands-
- cd /lib/x86_64-linux-gnu
- sudo ln -s libssl.so.1.0.0 libssl.so.10
-  sudo ln -s libcrypto.so.1.0.0 libcrypto.so.10

The steps for installation are-
1. sudo apt-get update
2. sudo apt-get install libssl1.0.0 libssl-dev
3. cd /lib/x86_64-linux-gnu
4. sudo ln -s libssl.so.1.0.0 libssl.so.10
5. sudo ln -s libcrypto.so.1.0.0 libcrypto.so.10
 The question is here now that do we need to put these libraries in activity itself because it is problem of ubuntu itself so can't be solved without re-installation libraries or creating symbolic links.

quozl commented:
The installation steps are not practical.  A child learner or elementary school teacher won't be able to follow those steps, because they are too complex, or because they require administration rights.  The x11vnc file is also very very old, so we can't use it.  See https://github.com/sugarlabs/VncLauncher/issues/7 for how this should be fixed.

Saumya-Mishra9129 committed:
Gdk.RGBA color parsing and removed jarabe.model.network

Saumya-Mishra9129 commented:
@quozl What vnc file we can use other than x11vnc file to remove #7 .I am not getting this. Do we need replacement of vnc file?

chimosky commented:
There's a checklist in #7, you can take a look at it.

To elaborate, use distro specific binaries as the ones used in the activity are old and have vulnerabilities.

quozl commented:
> @quozl What vnc file we can use other than x11vnc file to remove #7 .I am not getting this. Do we need replacement of vnc file?

Sorry, I did think it was obvious, and surprised I have to explain.  :grin:  Rephrasing part of #7, `/usr/bin/x11vnc` is available on Linux as a package, so assume it to be present as a system dependency.

It is much the same reason why we don't include `/usr/bin/python` in this repository.  We _could_ include python, but then we'd be running an old version of python fairly quickly.

In future, strategic plans for an issue are best discussed in the issue; https://github.com/sugarlabs/VncLauncher/issues/7, otherwise we fragment the discussion.

Saumya-Mishra9129 commented:
@quozl If you want to merge , you can merge this pr , I will create a new one for #7 , as it is difficult to find all related dependencies related to x11vnc. Thanks

chimosky commented:
> @quozl If you want to merge , you can merge this pr , I will create a new one for #7 , as it is difficult to find all related dependencies related to x11vnc. Thanks

I think the fix for #7 needs to be incorporated here so we can know if the activity works as intended and why are you looking for all dependencies related to x11vnc if I may ask?

The idea behind #7 is using a distro specific binary when available, [pkgs.org](https://pkgs.org/search/?q=x11vnc) lists the `x11vnc` package name for various distros. 

You'll want to;
- [ ] Install `x11vnc` in your test environment.
- [ ] Make the necessary changes .ie. remove support for the use of embedded binaries and use distro package if present.
- [ ] add a warning to advise of any missing dependency when the activity is started, if `x11vnc` isn't installed. 
- [ ] add dependency for Fedora and Ubuntu to README.md, (see other activities with detailed dependencies for examples)

The [x11vnc homepage](http://www.karlrunge.com/x11vnc/) might be of help.

quozl commented:
I agree, it should be straightforward, and you don't need to know the dependencies of x11vnc as these will be handled automatically by whatever package management system is used.

It may help to understand why this situation happened.  The activity was written for the OLPC XO laptop.  Activities could not depend on the ability to install dependencies, because they did not run as root, and there was an amazing security system called Bitfrost.  So activities that needed dependencies ended up with embedded binaries, which complicated their maintenance but got the job done.

Now, the activities don't have to focus on the same environment, and we can assume that the user has the capacity to install dependencies if needed.

This does mean that any user who downloads the .xo bundle from a Sugar app store won't necessarily be able to install the dependency, which is why the warning is needed.

Saumya-Mishra9129 commented:
@chimosky @quozl Thanks for your suggestions. Now I am thinking to include a script related to installation of x11vnc which must work on all distros possible.. and I will add steps for running the script in Readme. Will It be good or not, You both suggest please.

chimosky commented:
I think adding a script for it is a bit overkill, you can just include
instructions for some distros in the Readme - there's an existing commit
you can look at for such -.

On Sat, May 2, 2020, 3:07 PM Saumya Mishra <notifications@github.com> wrote:

> @chimosky <https://github.com/chimosky> @quozl <https://github.com/quozl>
> Thanks for your suggestions. Now I am thinking to include a script related
> to installation of x11vnc which must work on all distros possible.. and I
> will add steps for running the script in Readme. Will It be good or not,
> You both suggest please.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sugarlabs/VncLauncher/pull/9#issuecomment-622959114>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD3UAOREJOCMKNCOKBROOTLRPQSJFANCNFSM4L2MM2ZA>
> .
>

quozl commented:
@Saumya-Mishra9129, thanks for the offer, but I don't think a script is needed.

Let me put this activity into perspective.

* it is for teachers to use,
* it shares the display of a Sugar system on a local network,
* it does not provide any authentication or password.

It's 2006.  When the OLPC XO was produced, classroom displays and projectors were rare and expensive, and usually were not available in the third-world global-south where we wanted to deploy the laptops.  So we did not include an external socket for connecting to them.  This saved cost, risks of a physical gap in the casing, risks of electrostatic discharge damage, risks of a VGA connector being used to bend the circuit board when someone tripped over a cable, and VGA connectors had a limited lifetime of several hundred insertion removal cycles. We also increased a focus on collaboration via wireless networking.  Sugar was part of OLPC XO product and was not easily available on other computers.

It's 2008.  The VNC Launcher activity was also created, by volunteers, to provide a way for teachers in first-world scenarios to use a projector with an attached computer.

It's 2020.  What has changed since then;

* Sugar can be run on ordinary computers, and demonstrated using a projector,
* most laptops now have external video connectors of some sort, (HDMI is reasonably ESD resistant and the connector would more likely break rather than transfer significant force to the laptop),
* cost of displays and projectors has fallen, by a factor or ten or more,
* risk of automated network attack has risen, along with a focus on encryption and security protocols,

I don't recommend spending much effort in maintaining this activity.  It is such a small corner case, and the people who need it will have the skills to follow instructions.  They won't need a script.  We also can't afford to _maintain_ a script against changes made to all distros.

Hope that helps!

Saumya-Mishra9129 commented:
> @Saumya-Mishra9129, thanks for the offer, but I don't think a script is needed.
Let me put this activity into perspective.
it is for teachers to use,
it shares the display of a Sugar system on a local network,
it does not provide any authentication or password.
It's 2006. When the OLPC XO was produced, classroom displays and projectors were rare and expensive, and usually were not available in the third-world global-south where we wanted to deploy the laptops. So we did not include an external socket for connecting to them. This saved cost, risks of a physical gap in the casing, risks of electrostatic discharge damage, risks of a VGA connector being used to bend the circuit board when someone tripped over a cable, and VGA connectors had a limited lifetime of several hundred insertion removal cycles. We also increased a focus on collaboration via wireless networking. Sugar was part of OLPC XO product and was not easily available on other computers.
It's 2008. The VNC Launcher activity was also created, by volunteers, to provide a way for teachers in first-world scenarios to use a projector with an attached computer.
It's 2020. What has changed since then;
Sugar can be run on ordinary computers, and demonstrated using a projector,
most laptops now have external video connectors of some sort, (HDMI is reasonably ESD resistant and the connector would more likely break rather than transfer significant force to the laptop),
cost of displays and projectors has fallen, by a factor or ten or more,
risk of automated network attack has risen, along with a focus on encryption and security protocols,
I don't recommend spending much effort in maintaining this activity. It is such a small corner case, and the people who need it will have the skills to follow instructions. They won't need a script. We also can't afford to maintain a script against changes made to all distros.
Hope that helps!

Yeah!! It will help for sure. Now I think including instructions for some distros in the Readme will be enough for it. Thanks

Saumya-Mishra9129 committed:
remove embedded x11vnc binaries

Saumya-Mishra9129 committed:
Add a warning for missing dependency

Saumya-Mishra9129 committed:
flake8 fix : E231 missing whitespace after ','

Saumya-Mishra9129 committed:
flake8 fix : F401 'platform' imported but unused

Saumya-Mishra9129 committed:
flake8 fix : E303 too many blank lines

Saumya-Mishra9129 committed:
flake8 fix : E301 expected 1 blank line, found 0

Saumya-Mishra9129 committed:
flake8 fix : E501 line too long (83 > 79 characters)

Saumya-Mishra9129 committed:
Add .flake8

Saumya-Mishra9129 commented:
@quozl @chimosky @pro-panda I have done with changes required for x11vnc , kindly review now. Done as suggested by @chimosky [here](https://github.com/sugarlabs/VncLauncher/pull/9#issuecomment-621971980).

Saumya-Mishra9129 commented:
Perhaps we can think here for releasing new version as specified in #1 . Also activity doesn't specify a license.

chimosky commented on README.md:
```diff
@@ -11,3 +11,32 @@ How to use?
 * [How to use VncLauncher](http://wiki.laptop.org/go/VNC_Launcher)
+
+
+Bridge depends on Python, [Sugar
```
Seems this was copy pasted without changing the activity name.

chimosky commented:
Thanks, reviewed 945f7b8.

Saumya-Mishra9129 commented:
Fixes #2 , #3 and #7

quozl commented:
Thanks.  Reviewed.

* [ ] the missing dependency warning must be visible to the user; something in logs 8492c69d340d5391f90ecda8c942743092082511 is not visible,
* [x] _how to use_ should be in README.md not on an old Wiki page on another project; we should no assume the user has internet,
* [x] _how to install dependencies_ says twice it is for development; it isn't for development only,
* [x] _how to develop_ is just normal, no need to have it,

I'm a bit worried about the method used to get an IPv4 address from the system.  The activity, being very old, used an old API, as does parts of Sugar still.  Now that we have ported to PyGObject, the NM module may be appropriate.  I'm unsure.

I'm a bit worried about the testing of getting an IPv4 address.  https://github.com/sugarlabs/sugar/commit/a9f6b994dfc87e4aa95744d49397977a548cae63 shows how we found the Ip4Address property to be unreliable.  Describe your testing, and if any of your test systems had multiple IP addresses?

quozl commented:
For interest, `_configure_vte` was copied from the Terminal activity a long time ago.  Since then, the Terminal activity has been refactored, and `configure_terminal` is the core of the functionality now.

Saumya-Mishra9129 commented:
>  the missing dependency warning must be visible to the user; something in logs 8492c69 is not visible,

When we click on Start button to start X11 VNC Server , after clicking enter, bash automatically gives error as `bash: /usr/bin/x11vnc: No such file or directory`. Do we need any other warning as well? I think one possible way can be - when user clicks Start  button a dialog window should come up telling to install library, if we could add this feature it can be more user friendly.

Saumya-Mishra9129 committed:
Use str as arg in configparser

configparser accepts <class 'str'> as *args.

Boolean and integer args had to be converted to str

Saumya-Mishra9129 committed:
Change method name : _configure_vte to configure_terminal

Saumya-Mishra9129 commented:
> I'm a bit worried about the method used to get an IPv4 address from the system. The activity, being very old, used an old API, as does parts of Sugar still. Now that we have ported to PyGObject, the NM module may be appropriate. I'm unsure.
> 
> I'm a bit worried about the testing of getting an IPv4 address. sugarlabs/sugar@a9f6b99 shows how we found the Ip4Address property to be unreliable. Describe your testing, and if any of your test systems had multiple IP addresses?

I tested it on VMWare with sugar live build 0.116 as server machine which had wired connection, and for client I used Remote Desktop viewer software on Ubuntu 19.10. I got it working fine . You can see [here](https://drive.google.com/file/d/1sO73l9dNZlSbTJbJaJBUXfKtdOC_JsX1/view?usp=sharing) also.

srevinsaju commented:
@Saumya-Mishra9129 I would suggest you to use any other platform for sharing videos / files. Google Drive is closed source and also resource intensive. Moreover, many reviewers prefer to use open source systems in development. In future, it might help; 
Moreover, these links are not static. A future contributor might not understand what this PR wanted to achieve, if you move the file in your Google Drive to trash or change the path. See the number of PRs on [sugarlabs/sugar](https://github.com/sugarlabs/sugar) which are incomplete because we have no clue on how to proceed as the images linked to the PR got removed. Thanks

> (I had told the others, in case you had not seen)

Saumya-Mishra9129 commented:
> @Saumya-Mishra9129 I would suggest you to use any other platform for sharing videos / files. Google Drive is closed source and also resource intensive. Moreover, many reviewers prefer to use open source systems in development. In future, it might help;
> Moreover, these links are not static. A future contributor might not understand what this PR wanted to achieve, if you move the file in your Google Drive to trash or change the path. See the number of PRs on sugarlabs/sugar which are incomplete because we have no clue on how to proceed as the images linked to the PR got removed. Thanks
> 
> (I had told the others, in case you had not seen)

I have shifted it to my repository so that anyone could see it [ here.](https://github.com/Saumya-Mishra9129/MyFiles/blob/master/vncLauncher.mp4)

quozl commented:
Thanks.  I've looked again at the way an IP address is determined;

https://github.com/sugarlabs/VncLauncher/blob/b9393801fc2c176ee54598f76030a2e319a0f59c/vnclauncher.py#L46-L66

Am I reading it right; it will iterate through the available network interfaces, ignore any of the local interfaces, and set the label of the button to show each IP, with the result that only the last of the IP addresses will be shown?

Activity was written for an OLPC XO, which had one network device, `wlan0`, which would have one IP address, either allocated through a wireless access point via DHCP, or by link-local addressing when using mesh or ad-hoc network mode.

Nowadays, there are laptops and computers with both wireless and wired network connections.

The Frame (F6) shows a device node for each network connection, and mouse-over or right-click shows IP address.  At the time this activity was written, Sugar probably did not do that well.

Saumya-Mishra9129 commented:
> Am I reading it right; it will iterate through the available network interfaces, ignore any of the local interfaces, and set the label of the button to show each IP, with the result that only the last of the IP addresses will be shown?
> 
> Activity was written for an OLPC XO, which had one network device, wlan0, which would have one IP address, either allocated through a wireless access point via DHCP, or by link-local addressing when using mesh or ad-hoc network mode.
> 
> Nowadays, there are laptops and computers with both wireless and wired network connections.
> 
> The Frame (F6) shows a device node for each network connection, and mouse-over or right-click shows IP address. At the time this activity was written, Sugar probably did not do that well.

Yeah , You are probably right, There must be some reason why author of the activity did that. One possible reason can be `def __get_devices_reply_cb(self, devices)` takes dbus.Array as a argument .

Saumya-Mishra9129 commented:
The work on this PR is almost completed, @quozl look if you can merge after testing.

quozl commented:
Shouldn't both IP addresses be shown in a situation with both wireless and wired connections active?

Saumya-Mishra9129 commented:
> Shouldn't both IP addresses be shown in a situation with both wireless and wired connections active?

Yes There should be, as it should depend on choice of user which connection wired or wireless he want to use. But probably that time it was designed for OLPC XO and it had only one network device, but now situation has changed. I am not that much familiar with how this can be done. I need to do some reasearch on this.

quozl commented:
Well, as I said, line 65 will be executed twice.  The second time it will overwrite the value set in the first place.

Saumya-Mishra9129 commented:
> Well, as I said, line 65 will be executed twice. The second time it will overwrite the value set in the first place.

Yeah I understood this. One possible way we can show all IP addresses to user. But I guess it can be complex as it is not necessary that user is familiar with IP addresses .

quozl commented:
I guess so.

Users of this activity are teachers, and the [instructions from 2011](http://wiki.laptop.org/go/VNC_Launcher) do give some indication.

If you think teachers won't be able to use IP addresses, then I guess the next thing to try is registering an mDNS name on the network and having the VNC client use that.

Saumya-Mishra9129 commented:
@JuiP Can you please look and plan for a new release of activity? 

> I guess so.
> 
> Users of this activity are teachers, and the instructions from 2011 do give some indication.
> 
> If you think teachers won't be able to use IP addresses, then I guess the next thing to try is registering an mDNS name on the network and having the VNC client use that.

Can we plan a release now, and leave it as I tried to find out the ways , I got to know about python3-avahi but it is not supported by all linux distributions, I couldn't find it for Ubuntu. 
Also can you look for a merge, so that @JuiP can start on new release.

quozl commented:
PyGObject has an Avahi API.  I've not used it.

Saumya-Mishra9129 commented:
> PyGObject has an Avahi API.  I've not used it.

Yeah I saw about this. But I have no experience with this and how this works. I am planning to work on this. But for now it would be out of scope of this pull request. I will make a new one probably after implementing Avahi.

quozl commented:
Please read the manual page for `x11vnc`, in particular the _-avahi_ option.

Tested.

* [x] the Vte does not have focus to begin with, so nothing can be done without clicking in the widget first,
* [x] exiting the Vte with ctrl+d or `exit` does not exit the activity; this is unlike Terminal; possibly the problem was fixed in Terminal long ago but the sources have diverged,
* [x] there is no message to inform the user that `/usr/bin/x11vnc` is not available; the button just inserts the command into the Vte without pressing enter, and when you press enter the shell reports _No such file or directory_,
* [ ] when in Sugar Ad-hoc network mode, the current IP address is not displayed, instead "Error!! check connection", workaround is to use Sugar Frame F6
* [x] README.md should advise use of `apt` not `apt-get`,
* [ ] in the default operating mode, `x11vnc` is used insecurely, related to but not mentioned in https://github.com/sugarlabs/VncLauncher/issues/7,
* [ ] the critical how to use information at http://wiki.laptop.org/go/VNC_Launcher should be in the README.md instead of being externally hosted,
* [x] no information is present on how to install a `vncviewer` (e.g. sudo apt install xtightvncviewer), and run it,

That said, it does work.

Saumya-Mishra9129 committed:
Convert string config args to parseable set_cursor_blink_mode

Fixes TypeError: Expected a Vte.CursorBlinkMode, but got str

Saumya-Mishra9129 committed:
Fix TypeError on set_scrollback

TypeError: Must be number, not str

Saumya-Mishra9129 committed:
Use str as args in configparser

configparser accepts str as argument. Boolean and integers values need to be converted to str first

Saumya-Mishra9129 committed:
Fix Vte Focus

 Vte didn't had focus to begin with. Workaround was to click in the widget first

Saumya-Mishra9129 commented:
Update - Please review till 05995ea. @chimosky @quozl .

quozl commented:
* [ ] scenario of two IP addresses (wireless and ethernet) still isn't handled; while we had discussed using Zeroconf/mDNS above, you decided against it, so that leaves the problem of having to handle the scenario,
* [x] the README.md doesn't show how to use the activity,
* [x] the README.md explains how to install a VNC viewer, but it isn't the packaged viewer available for the operating system.

Re: my earlier comment https://github.com/sugarlabs/VncLauncher/pull/9#issuecomment-608295946 ... I still don't see this winning any users.  As far as I can see from the limited information; a teacher would have to install this activity on her device, install x11vnc on her device, then for each child device install a VNC viewer, and teach the child how to type a Terminal command.  Where's the corresponding viewer activity?  This [search](http://activities.sugarlabs.org/en-US/sugar/search?q=vnc&cat=all) is interesting; there's a viewer that isn't on GitHub yet.

(Also, for history, one of the reasons the activity existed was because the OLPC XO had no external monitor socket, and teachers with classroom projectors needed a way to show the Sugar desktop from an XO using their Windows classroom system.  They would install this activity on the XO, a VNC viewer on their Windows system, and demo from there.  The method we eventually settled on at OLPC was to sell USB to VGA adapters.  Now, the latest OLPC hardware has external monitor sockets.)

Saumya-Mishra9129 commented:
> I still don't see this winning any users. As far as I can see from the limited information; a teacher would have to install this activity on her device, install x11vnc on her device, then for each child device install a VNC viewer, and teach the child how to type a Terminal command.

In 51344d9, I tried to give a feature by which when a user installs activity for the first time, he/she will be automatically able to install x11vnc without a need to type terminal commands. I tested it with Ubuntu and Debian, If @chimosky can test this with fedora then it will be good to go. 

> Where's the corresponding viewer activity? This search is interesting; there's a viewer that isn't on GitHub yet.

Thanks I saw this , but no information where source code can be found. Should we try to contact to Author of the **Xo VNC Viewer**. ?

> scenario of two IP addresses (wireless and ethernet) still isn't handled; while we had discussed using Zeroconf/mDNS above, you decided against it, so that leaves the problem of having to handle the scenario,

I couldn't find a way to use x11vnc 's avahi option. I am trying to find out a way.

> the README.md explains how to install a VNC viewer, but it isn't the packaged viewer available for the operating system.

Actually I think it is a easy way to install RealVNC this way. Setting up TightVNC with terminal is so complex , It requires a large setup. Downloading this way can work on Windows and Linux both as you mentioned about windows also above.

chimosky commented on vnclauncher.py:
```diff
@@ -106,14 +106,18 @@ def __init__(self, handle):
         self._vte.grab_focus()
 
+    def _quit_cb(self, vte, data=None):
+        print("Quitting...")
```
Why print? when logging calls were used in the activity.

Saumya-Mishra9129 replied:
Removed.

chimosky commented on vnclauncher.py:
```diff
@@ -128,9 +129,28 @@ def connectVNC(self, button):
+
+            if response == Gtk.ResponseType.OK:
+                logging.debug("Installing x11vnc")
+                if platform.version().find("Ubuntu") > -1 or platform.version().find("Debian"):
+                    cmd = "apt install x11vnc"
+                if platform.version().find("Fedora") > -1:
```
Might be better to use `platform.linux_distribution()[0]` because on fedora systems, `platform.version().find("Fedora")` will always return `-1` as `Fedora` isn't part of the string returned by `platform.version()`. Although it's deprecated in python>=3.5.

chimosky replied:
Find a way to check for `Fedora` distros.

chimosky replied:
`platform.platform()` contains the necessary information.

Saumya-Mishra9129 replied:
Thanks Done in c733b9b.

chimosky commented:
@quozl said
>Re: my earlier comment #9 (comment) ... I still don't see this winning any users. As far as I can see from the limited information; a teacher would have to install this activity on her device, install x11vnc on her device, then for each child device install a VNC viewer, and teach the child how to type a Terminal command. 

I agree.

quozl commented:
@Saumya-Mishra9129 wrote:
> Should we try to contact to Author of the **Xo VNC Viewer**. ?

What for?  We have the activity bundle, isn't that enough to begin with?

An ideal VNC activity would offer a "share my desktop" button, support Sugar Toolkit collaboration, and joiners would automatically start a VNC viewer, perhaps in read-only mode.  Obtaining the IP address through the collaboration stack is fairly easy, see the Level activity for details.

Saumya-Mishra9129 committed:
Update README.md

- add x11vnc dependency for Fedora , Ubuntu and Debian.
- add info required to install VNC Viewer.
- Modify how to use section

Saumya-Mishra9129 committed:
Fix Exit of Vte with ctrl+d or exit

Saumya-Mishra9129 committed:
Add Message Dialog to install x11vnc

Saumya-Mishra9129 commented:
> What for? We have the activity bundle, isn't that enough to begin with?.

Yeah It is. Thanks. I will try to look at the bundle and will what maintenance required there.

@quozl 
> the README.md doesn't show how to use the activity,

This I have tried to do in 64df965. You can review.

quozl commented:
Thanks.  Updated checkboxes.

Saumya-Mishra9129 committed:
fix multiple IP addresses issue

 The activity was written for an OLPC XO years ago, which had one network device wlan0, which would have one IP address. Nowadays, We use sugar on laptops and computers with both wireless and wired network connections.
 The patch displays all the IP(IPv4 and IPV6) addresses assigned to your system (Wired and Wireless)

Saumya-Mishra9129 commented:
> fix multiple IP addresses issue 477d141, 
 
@quozl Please review and if possible merge the changes. I have tested on Ubuntu and Debian works fine for both systems.
It looks like this -

![Screenshot from 2020-08-13 16-35-53](https://user-images.githubusercontent.com/45252461/90131084-28069700-dd89-11ea-8125-6450e6722dc3.png)
